src/openms/source/QC/DBSuitability.cpp

```diff
@@ -55,58 +66,419 @@ namespace OpenMS
+      throw Exception::InternalToolError(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "Return state was: " + static_cast<Int>(indexer_exit));
+    }
+
+    if (keep_files)
+    {
+      id_file.store(tmp_dir.getPath() + "indexed_pre_FDR.idXML", prot_ids, pep_ids);
+    }
+
+    // calculate q-values
+    FalseDiscoveryRate fdr;
+    Param p(fdr.getParameters());
+    if (!p.exists("use_all_hits") || !p.exists("add_decoy_peptides") || !p.exists("add_decoy_proteins"))
+    {
+      throw Exception::InvalidParameter(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "FDR parameters probably changed. 'use_all_hits', 'add_decoy_peptides' or 'add_decoy_proteins' not found.");
+    }
     p.setValue("use_all_hits", "true");
     p.setValue("add_decoy_peptides", "true");
     p.setValue("add_decoy_proteins", "true");
 
-    FalseDiscoveryRate fdr;
     fdr.setParameters(p);
+    OPENMS_LOG_DEBUG << "Calculating q-values ..." << endl << endl;
     fdr.apply(pep_ids);
 
-    for (PeptideIdentification& pep_id : pep_ids)
+    if (keep_files)
     {
-      // sort hits by q-value
-      pep_id.sort();
+      id_file.store(tmp_dir.getPath() + "indexed_post_FDR.idXML", prot_ids, pep_ids);
+    }
 
-      vector<PeptideHit>& hits = pep_id.getHits();
+    return pep_ids;
+  }
+
+  std::vector<FASTAFile::FASTAEntry> DBSuitability::getSubsampledFasta_(const std::vector<FASTAFile::FASTAEntry>& fasta_data, double subsampling_rate) const
+  {
+    if (subsampling_rate < 0 || subsampling_rate > 1)
+    {
+      throw Exception::IllegalArgument(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "Subsampling rate has to be between 0 and 1. Aborting!");
+    }
+    Size num_AA{};
+    for (const auto& entry : fasta_data)
+    {
+      num_AA += entry.sequence.size();
+    }
+    double num_AS_written = num_AA * subsampling_rate;
+
+    Math::RandomShuffler shuffler;
+    shuffler.seed(UniqueIdGenerator::getUniqueId());
+    std::vector<int> rnd_indices(fasta_data.size());
+    std::iota(std::begin(rnd_indices), std::end(rnd_indices), 0);
+    shuffler.portable_random_shuffle(rnd_indices.begin(), rnd_indices.end());
+
+    Size curr_AA{};
+    vector<FASTAFile::FASTAEntry> sampled_fasta;
+    for (const int i : rnd_indices)
+    {
+      if (curr_AA >= num_AS_written) break;
+      sampled_fasta.push_back(fasta_data[i]);
+      curr_AA += fasta_data[i].sequence.size();
+    }
+    return sampled_fasta;
+  }
+
+  void DBSuitability::calculateSuitability_(const std::vector<PeptideIdentification>& pep_ids, SuitabilityData& data) const
+  {
+    bool no_re_rank = param_.getValue("no_rerank").toBool();
+    double cut_off_fract = param_.getValue("reranking_cutoff_percentile");
+    double FDR = param_.getValue("FDR");
+
+    if (pep_ids.empty())
+    {
+      OPENMS_LOG_WARN << "No peptide identifications found in given idXML! No calculations performed." << endl;
+      return;
+    }
+
+    if (!no_re_rank)
+    {
+      data.cut_off = getDecoyCutOff_(pep_ids, cut_off_fract);
```

Comment:
setting `data.cutoff` (or any other member of `data`) is a bit dangerous, since you never know what `data` was previously populated with (maybe the user passed in an old object?).
Better create a fresh `SuitabilityData` locally, and return it.
Or was there a specific reason to do it this way?

Reply:
This is done because an empty `SuitabilityData` object is pushed into the results vector at the very beginning and all changes are made on a reference to that object.
I guess I could implement a `clear()` function and clear the object at the beginning of `calculateSuitability_`.